src/lib/aggregator.ts

```diff
@@ -1,58 +1,144 @@
- * Otherwise iterates over inputs by aggregating per given `metrics`.
- */
-export const aggregate = (
-  inputs: PluginParams[],
-  metrics: string[],
-  parameters: Parameters
-) => {
-  checkIfMetricsAreValid(metrics, parameters);
-
-  return inputs.reduce((acc, input, index) => {
-    for (const metric of metrics) {
-      if (!(metric in input)) {
-        throw new InvalidAggregationParams(METRIC_MISSING(metric, index));
+/**Returns Boolean indicating whether a node has children */
+function hasChildren(node: Node) {
+  if (node['children']) {
+    return true;
+  }
+  return false;
+}
+
+/** Returns children of given node */
+function getChildren(node: Node): Node[] {
+  const children: Node[] = [];
+  Object.keys(node.children).forEach(child => {
+    children.push(node.children[child]);
+  });
+  return children;
+}
+
+/** element-wise addition of two arrays */
+function aggregateArrays(a: Array<any>, b: Array<any>, method: string) {
+  if (method === 'sum') {
+    return a.map((element, idx) => {
+      return element + b[idx];
+    });
+  } else if (method === 'avg') {
+    return a.map((element, idx) => {
+      return (element + b[idx]) / 2;
+    });
+  }
+  return a;
+}
+
+/** aggregates multiple time series into one single time series*/
+export const aggregate = (tree: Node, context: ManifestCommon) => {
+  checkIfMetricsAreValid(context);
+  // check if root has children
+  if (hasChildren(tree)) {
+    // iterate through child nodes
+    getChildren(tree).forEach(node => {
+      // if the child has children, start recursive call to aggregate()
+      if (hasChildren(node)) {
+        aggregate(node, context);
       }
+      context.aggregation?.metrics.forEach(metric => {
+        let treeTimeSeries = tree.inputs?.map(input => input[`${metric}`]);
 
-      const accessKey = `aggregated-${metric}`;
-      acc[accessKey] = acc[accessKey] ?? 0;
-      acc[accessKey] += parseFloat(input[metric]);
+        if (
+          treeTimeSeries === undefined ||
+          treeTimeSeries?.includes(undefined)
+        ) {
+          treeTimeSeries = new Array(3).fill(0);
+        }
+
+        let nodeTimeSeries = node.inputs?.map(input => input[`${metric}`]);
 
-      /** Checks for the last iteration. */
-      if (index === inputs.length - 1) {
-        if (getAggregationMethod(metric, parameters) === 'avg') {
-          acc[accessKey] /= inputs.length;
+        if (
+          nodeTimeSeries === undefined ||
+          nodeTimeSeries?.includes(undefined)
+        ) {
+          nodeTimeSeries = new Array(3).fill(0);
         }
-      }
-    }
 
-    return acc;
-  }, {} as AggregationResult);
+        if (
+          context.aggregation?.type === 'both' ||
+          context.aggregation?.type === 'vertical'
+        ) {
+          if (tree[`aggregated-${metric}`]) {
+            if (node[`aggregated-${metric}`]) {
+              tree[`aggregated-${metric}`] = aggregateArrays(
```

Comment:
I _think_ what we are doing here would only work if a parent node had 1 child since the aggregateArrays only aggregates two arrays together.

More specifically i think this logic only works if the number of children is > 1 AND the aggregation method is sum since it will just keep on adding the values to the parent node.
OR
If the aggregation method is average it would only work if there is 1 child since the seconds childs average would somehow overwrite the average figure from the first child, i don't think this would average all the children values and put them on the parent.

We also need to support other aggregation options, I can imagine scenarios where we just pick the first one in the list, or maybe do some string manipulation if you are aggregating a descriptive field. Unsure, but it only really works I _think_ if you are aggregating all the children in one go with each other so you know all the values.